File: Max Circular Subarray Sum.py
def circularSubarraySum(arr):
    # Compute total sum of the array
    total_sum = sum(arr)

    # Helper function to calculate max or min subarray sum using Kadane's Algorithm
    def kadane(arr, find_max):
        running, result = 0, float('-inf') if find_max else float('inf')
        for e in arr:
            running = (max if find_max else min)(e, running + e)
            result = (max if find_max else min)(result, running)
        return result

    # Maximum subarray sum (non-circular)
    max_subarray_sum = kadane(arr, True)

    # Minimum subarray sum (to find the worst-case wrap-around sum)
    min_subarray_sum = kadane(arr, False)

    # If all elements are negative, return the maximum subarray sum (non-circular)
    if total_sum == min_subarray_sum:
        return max_subarray_sum

    # Maximum of non-circular and circular subarray sums
    return max(max_subarray_sum, total_sum - min_subarray_sum)

File: Max Circular Subarray Sum.cpp
class Solution {
public:
    int maxCircularSum(vector<int> &arr) {
        int n=arr.size();
        int totalSum=accumulate(arr.begin(),arr.end(),0);
        int currMax=arr[0],maxSum=arr[0];
        int currMin=arr[0],minSum=arr[0];
        for(int i=1;i<n;i++){
            currMax=max(arr[i],currMax+arr[i]);
            maxSum=max(maxSum,currMax);
            currMin=min(arr[i],currMin+arr[i]);
            minSum=min(minSum,currMin);
        }
        if(maxSum<0) return maxSum;
        return max(maxSum,totalSum-minSum);
    }
};
